### PR TITLE
DietPi-Software | CUPS

### LINKED ISSUE #3942: DietPi-Software | Add CUPS printer server
Hey there,

I was looking for a way to print to a PDF from my iPhone. I made it with the Raspberry Pi and Dietpi without any problems:

First installing CUPS:

```
sudo apt install avahi-daemon cups python3-cups
cupsctl --remote-admin 
cupsctl --share-printers
```
The CUPS-Webinterface is listening on Port 631

Now it is important to create virtual PDF-Printer. There is a package for it, called `printer-driver-cups-pdf` - but this printer sucks (prints images in a poor resolution)! It's better to create a General Printer from scratch:

Stop the Service `systemctl stop cups.services`

`nano /etc/cups/pdf-printer.conf`

Copy & Paste straight forward ...

**Important:** the Output-Dir of the PDFs: `AnonDirName /var/spool/cups-pdf/ANONYMOUS`

```
#  cups-pdf.conf -- CUPS Backend Configuration (version 3.0.1, 2017-02-24)
#  18.09.2005, Volker C. Behr
#  volker@cups-pdf.de
#  http://www.cups-pdf.de
#
#
#  This code may be freely distributed as long as this header
#  is preserved. Changes to the code should be clearly indicated.
#
#  This code is distributed under the GPL.
#  (http://www.gnu.org/copyleft/gpl.html)
#
#  For more detailed licensing information see cups-pdf.c in the
#  corresponding version number.

###########################################################################
#                                                                         #
# This is the configuration file for CUPS-PDF. Values that are not set in #
# here will use the defaults. Changes take effect immediately without the #
# need for restarting any services.                                       #
#                                                                         #
# Take care not to add whitespaces at the end of a line!                  #
#                                                                         #
# Options are marked where they can be set (setting via PPD requires the  #
# PPD file that comes with CUPS-PDF to be used!).                         #
# Options passed via lpoptions that are not named for lpoptions here are  #
# ignored.                                                                #
# Options precedence is as follows:                                       #
# 1st: lpoptions                                                          #
# 2nd: PPD settings                                                       #
# 3rd: config file (this file)                                            #
# 4th: default values                                                     #
###########################################################################


###########################################################################
#                                                                         #
# Path Settings                                                           #
#                                                                         #
###########################################################################

### Key: Out (config)
##  CUPS-PDF output directory
##  special qualifiers:
##     ${HOME} will be expanded to the user's home directory
##     ${USER} will be expanded to the user name
##  in case it is an NFS export make sure it is exported without
##  root_squash!
##  on Ubuntu, the cupsd AppArmor profile needs to be updated
##  to match the output path as per instructions in LP:147551
### Default: /var/spool/cups-pdf/${USER}

Out ${HOME}/PDF

### Key: AnonDirName (config)
##  ABSOLUTE path for anonymously created PDF files
##  if anonymous access is disabled this setting has no effect
### Default: /var/spool/cups-pdf/ANONYMOUS

AnonDirName /var/spool/cups-pdf/ANONYMOUS

### Key: Spool (config)
##  CUPS-PDF spool directory - make sure there is no user 'SPOOL' on your
##  system or change the path
### Default: /var/spool/cups-pdf/SPOOL

#Spool /var/spool/cups-pdf/SPOOL


###########################################################################
#                                                                         #
# Filename Settings                                                       #
#                                                                         #
###########################################################################

### Key: Truncate (config, ppd, lpoptions)
##  truncate long filenames to a maximum of <Truncate> characters
##  this does not consider the full path to the output but only the filename
##  without the .pdf-extension or a job-id prefix (see 'Label')
##  the minimal value is 8
### Default: 64

#Truncate 64

### Key: Cut (config, lpoptions)
##  removing file name extensions before appending .pdf to output
##  extensions will only be removed if _both_ the following criteria are met:
##   - the extension (w/o the dot) is not longer than <Cut> characters
##   - the remaining filename has a minimal length of 1 character
##  set Cut to -1 in order to disable cutting
##  recommended values: pure UNIX environment : -1
##                      mixed environments    :  3
### Default: 3

#Cut 3

### Key: Label (config, ppd, lpoptions)
##  label all jobs with a unique job-id in order to avoid overwriting old
##  files in case new ones with identical names are created; always true for
##  untitled documents
##  0: label untitled documents only
##  1: label all documents with a preceeding "job_#-"
##  2: label all documents with a tailing "-job_#"
### Default: 0

#Label 0

### Key: TitlePref (config, ppd, lpoptions)
##  where to look first for a title when creating the output filename
##  (title in PS file or title on commandline):
##  0: prefer title from %Title statement in the PS file
##  1: prefer title passed via commandline
### Default: 0

#TitlePref 0


###########################################################################
#                                                                         #
# User Settings                                                           #
#                                                                         #
###########################################################################

### Key: AnonUser (config)
##  uid for anonymous PDF creation (this might be a security issue)
##  this setting has no influence on AnonDirName (see there)
##  set this to an empty value to disable anonymous
### Default: nobody

#AnonUser nobody

### Key: LowerCase (config)
##  This options allows to check user names given to CUPS-PDF additionally
##  against their lower case variants. This is necessary since in some
##  Windows environments only upper case user names are passed. Usually UNIX
##  user names are all lower case and it is save to use this option
##  but be aware that it can lead to mis-identifications in case
##  you have user names that differ only in upper/lower case.
##     check only against user name as passed to CUPS  : 0
##     check additionally against lower case user name : 1
### Default: 1

#LowerCase 1

### Key: UserPrefix (config)
##  some installations require a domain prefix added to the user name
##  leave empty for no prefix
### Default: <empty>

#UserPrefix

### Key: DirPrefix (config)
##  if a prefix was defined above this switch toggels whether to include
##  the prefix in the output directory's name (if not $HOME) or not
##  0: do not include, 1: include
### Default: 0

#DirPrefix 0

### Key: RemovePrefix (config)
##  some installation pass usernames with a prefix (usually a domain name)
##  if you do not want this prefix to be used by the ${USER} variable for
##  output directories put the part which is to be cut here
### Default: <empty>

#RemovePrefix


###########################################################################
#                                                                         #
# Security Settings                                                       #
#                                                                         #
###########################################################################

### Key: AnonUMask (config)
##  umask for anonymous output
##  these are the _inverse_ permissions to be granted
### Default: 0000

#AnonUMask 0000

### Key: UserUMask (config, lptoptions)
##  umask for user output of known users
##  changing this can introduce security leaks if confidential
##  information is processed!
### Default: 0077

#UserUMask 0077

### Key: Grp (config)
##  group cups-pdf is supposed to run as - this will also be the gid for all
##  created directories and log files
### Default: lp

Grp lpadmin

### Key: AllowUnsafeOptions (config)
##  DON'T CHANGE THIS SETTING UNLESS YOU ABSOLUTELY KNOW WHAT YOU ARE DOING
##  set to 1 in order to allow users to override any option - including
##  those that pose SEVERE SECURITY RISKS, set to 0 for full security
### Default: 0

#AllowUnsafeOptions 0


###########################################################################
#                                                                         #
# Log Settings                                                            #
#                                                                         #
###########################################################################

### Key: Log (config)
##  CUPS-PDF log directory
##  set this to an empty value to disable all logging
### Default: /var/log/cups

#Log /var/log/cups

### Key: LogType (config, ppd)
##  log-mode
##  1: errors
##  2: status (i.e. activity)
##  4: debug - this will generate a lot of log-output!
##  add up values to combine options, i.e. 7 is full logging
##  if logging is disabled these setting have no effect
### Default: 3

#LogType 3


###########################################################################
#                                                                         #
# PDF Conversion Settings                                                 #
#                                                                         #
###########################################################################

### Key: GhostScript (config)
##  location of GhostScript binary (gs)
##  MacOSX: for using pstopdf (recommended) set this to /usr/bin/pstopdf
##          or its proper location on your system
### Default: /usr/bin/gs

#GhostScript /usr/bin/gs

### Key: GSTmp (config)
##  location of temporary files during GhostScript operation
##  this must be user-writable like /var/tmp or /tmp !
### Default: /var/tmp

#GSTmp /var/tmp

### Key: GSCall (config)
## command line for calling GhostScript (!!! DO NOT USE NEWLINES !!!)
## MacOSX: for using pstopdf set this to %s %s -o %s %s
### Default: %s -q -dCompatibilityLevel=%s -dNOPAUSE -dBATCH -dSAFER -sDEVICE=pdfwrite -sOutputFile="%s" -dAutoRotatePages=/PageByPage -dAutoFilterColorImages=false -dColorImageFilter=/FlateEncode -dPDFSETTINGS=/prepress -c .setpdfwrite -f %s

#GSCall %s -q -dCompatibilityLevel=%s -dNOPAUSE -dBATCH -dSAFER -sDEVICE=pdfwrite -sOutputFile="%s" -dAutoRotatePages=/PageByPage -dAutoFilterColorImages=false -dColorImageFilter=/FlateEncode -dPDFSETTINGS=/prepress -c .setpdfwrite -f %s

### Key: PDFVer (config, ppd, lptopions)
##  PDF version to be created - can be "1.5", "1.4", "1.3" or "1.2"
##  MacOSX: for using pstopdf set this to an empty value
### Default: 1.4

#PDFVer 1.4

### Key: PostProcessing (config, lptoptions)
##  postprocessing script that will be called after the creation of the PDF
##  as arguments the filename of the PDF, the username as determined by
##  CUPS-PDF and the one as given to CUPS-PDF will be passed
##  the script will be called with user privileges
##  set this to an empty value to use no postprocessing
### Default: <empty>

#PostProcessing


###########################################################################
#                                                                         #
# Experimental Settings                                                   #
#   These settings activate experimental options. If you decide to use    #
#   them I would appreciate any feedback - including an 'ok' if they      #
#   work as expected - so I can eventually put them into the non-         #
#   experimental sections.                                                #
#                                                                         #
###########################################################################

### Key: DecodeHexStrings (config)
##  this option will try to decode hex strings in the title to allow
##  internationalized titles
##  (have a look at pstitleconv on www.cups-pdf.de for a suitable filter
##   for data from Windows clients)
##  0: disable, 1: enable
### Default: 0

DecodeHexStrings 1

### Key: FixNewlines (config)
##  this option will try to fix various unusal line delimiters (e.g.
##  form feeds)
##  especially useful when using non-Linux-generated files
##  0: disable, 1: enable
### Default: 0

#FixNewlines 0
```
Replacing the `/etc/cups/printers.conf` File:

```
# Printer configuration file for CUPS v2.2.10
# Written by cupsd
# DO NOT EDIT THIS FILE WHEN CUPSD IS RUNNING
<DefaultPrinter PDF>
UUID urn:uuid:eb0decc5-b6f5-38dc-46e7-a001c26b49d1
Info PDF
MakeModel Generic CUPS-PDF Printer (w/ options)
DeviceURI cups-pdf:/
State Idle
StateTime 1606057228
ConfigTime 1606056136
Type 8450124
Accepting Yes
Shared Yes
JobSheets none none
QuotaPeriod 0
PageLimit 0
KLimit 0
OpPolicy default
ErrorPolicy retry-job
Option pdftops-renderer pdftocairo
</DefaultPrinter>
```

Restart the Services `systemctl restart cups.services`

Then getting a Script which generates the Airprint-Deamon
```
wget https://raw.githubusercontent.com/tjfontaine/airprint-generate/master/airprint-generate.py
chmod 755 airprint-generate.py
mv AirPrint-*.service /etc/avahi/services
./airprint-generate.py -d /etc/avahi/services
```
Restart the Services `systemctl restart avahi-daemon.service`

Voilá:
image

### PR DESCRIPTION
**Pull request by Joulinar**
CUPS new software title (ID 187)

**Status**: Ready
- [x] add new software title CUPS
- [x] add new service CUPS
- [x] add change log
- [x] add `avahi-daemon` support

**Testing**: install/reinstall/uninstall/printing
- [x] Buster RPi1 32bit
- [x] Buster RPi3B+ 32bit
- [x] Buster RPi4B 64bit
- [x] Bullseye VM
- [x] Buster VM
- [x] Stretch VM
- [x] test print on W10
- [x] test print on macOS
- [x] test print on Android
- [x] test print on iOS

**Reference**: https://github.com/MichaIng/DietPi/issues/3942

**Commit list/description**:
- DietPi-Software | CUPS: add new software title
- DietPi-Services | CUPS: add new software title
- DietPi-Software | CUPS: add `avahi-daemon` as dependency
- DietPi-Survey_report | Add CUPS support, newly added with DietPi v7.1
- DietPi-Software | CUPS - Download base configuration if it does not exist yet

### CODE REVIEW

**Comment by Joulinar** [comment]
ok first attempt, with quite a simple installation procedure.

**Comment by Joulinar** [comment]
at the moment I'm puzzled, struggling with printing on Android nativ printing functionality. There is no issue to detect printers on W10 as well as iOS. However Android is strange. I have 3 devices, a S8, S9 and a Pixel 4. While the S9 is working fine, printers are not detected on the S8 + Pixel 4 🤔

**Comment by MichaIng** [comment]
Good that everything else works already. I'll test it on my Samsung A3 with Android 8.0 and two printers around here.

File: dietpi/dietpi-software [hunk 1]
```diff
@@ -1880,6 +1880,14 @@ _EOF_
 		aSOFTWARE_CATEGORY_INDEX[$software_id]=18
 		aSOFTWARE_ONLINEDOC_URL[$software_id]='https://dietpi.com/docs/software/printing/#octoprint'
 		(( $G_HW_ARCH < 3 && $G_DISTRO > 4 )) || aSOFTWARE_REQUIRES_BUILDESSENTIAL[$software_id]=1
+		#------------------
+		software_id=187
+
+		aSOFTWARE_NAME[$software_id]='CUPS'
+		aSOFTWARE_DESC[$software_id]='common UNIX Pprinting system'
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Typo: `Pprinting` => `printing`

**Comment by Joulinar** [comment]
@MichaIng 
One challenge, on reinstall we will apply local access configuration again. Is this ok or should we skip this step on reinstall?

**Comment by MichaIng** [comment]
IMO we should leave custom configs untouched. The the package ships with a default config file, either we could set a variable prior to package install or check for some content that only exists on a new config file.

**Comment by Joulinar** [comment]
by default LISTEN port is ~~0.0.0.0~~ `localhost`. We could assume that this is valid for new/untouched configs. Because usually people will change this to be able to access the CUPS 😉

**Comment by MichaIng** [comment]
I see `Listen localhost:631`, so you mean `127.0.0.1`, I guess?

**Comment by Joulinar** [comment]
ah sorry, my fault 😃 Yes it is `localhost` by default

**Comment by MichaIng** [comment]
Not 100% failsafe, since it could be setup with reverse proxy like this, I guess?

**Comment by Joulinar** [comment]
let me check the behaviour if config file exist before apt package installation

**Comment by Joulinar** [comment]
ok an existing config will not be overwritten by the apt package installation. Means we could do it as follow

1. on an empty system check for availability of `cupsd.conf` 
2. if not present, pull our own `cupsd.conf` config file 
3. perform `apt install cups`

on reinstall
1. system check for availability of `cupsd.conf` 
2. if present, do not pull `cupsd.conf` config file
3. perform `apt reinstall cups`

File: dietpi/dietpi-software (hunk 1 above)
Comment (Joulinar) [reply dietpi/dietpi-software]: will be updated once we agreed on installation procedure

**Comment by MichaIng** [comment]
🈯 Works pretty fine on Buster VM, including web UI, adding my two test printers (USB pass-through), finding them on Windows 10 as network printers, printing test pages from Windows 10 and from web UI.

Now testing on Android 8, using the FOSS CUPS print Android app...
🈯 ... both printers are found as network printers.
🈯 ... I can't test actual printing Android as I freezed the print spooler app and I cannot unfreeze it without unfreezing all other bloat, as the related app got broken with last Android update 😄.

Testing on macOS...
🈯 ... both printers were found as network printers.
🈯 ... printing works fine 😃.
___________
🈯 Same on Debian Bullseye.
___________
Tested web access and permissions:
- Login works via every UNIX user
- Printing via web UI is only possible for members of `lp` group, e.g.: `usermod -aG lp dietpi`
- Adding/managing printers via web UI is only possible for members of `lpadmin` group, e.g.: `usermod -aG lpadmin dietpi`
- When managing printers, the access is automatically upgraded to HTTPS. At least with current Chromium-based browsers, this is not a problem since the missing trusted certificate warning can be ignored one time and that decision is remembered.
- So all works as expected regarding access permissions.
___________
🈯 Same on Debian Bullseye.
- The web UI does not show the login prompt immediately, but only when trying to add/manage printers. But that's fine, IMO.

**Comment by Joulinar** [comment]
1. yes HTTPS is forced by CUPS if you like to access the admin area.
2. I will add `dietpi` group to user `lp` and `lpadmin`
3. yes login is needed on admin page only if you like to configure stuff. This is due to setting `Require user @SYSTEM` on the config file for these areas. If needed we could add this to the initial page as well

**Comment by nonnorm** [comment]
I can verify that installing and adding a printer works on RPi 4B 64-bit.  I'll test printing with iOS later when I get home.

**Comment by MichaIng** [comment]
> I will add `dietpi` group to user `lp` and `lpadmin`

You mean add `dietpi` user to `lp` and `lpadmin` groups? No please skip that. I don't want to add any further steps to make `dietpi` a required login user, instead I want to remove it and allow creating a custom user on first boot instead. Having that user where not all "users" know about, which can be used to login via SSH etc is a security issue. It's fine for now that OOTB root login works, and if users use any other login user, reasonably, then they need to add it to `lpadmin`, which is fine when documented.

An alternative I could think if is to create a new `cups` user that is explicitly created for CUPS web UI login, and CUPS could be configured to allow only that user and no other. But since it needs to be a login user with password, it can be used to login via SSH as well, so it would need to be isolated, disabling the login shell and use a `/nonexistent` home directory to not cause another attack vector 😉.

> yes login is needed on admin page only if you like to configure stuff. This is due to setting `Require user @SYSTEM` on the config file for these areas. If needed we could add this to the initial page as well

Ah, probably I did remember it from Buster and Bullseye tests wrong. It's fine IMO.

**Comment by Joulinar** [comment]
ok in this case we need to finalize testing only and maybe check out Android

**Comment by MichaIng** [comment]
> and maybe check out Android

As said, while detecting the printers works pretty fine, I broke the printing spooler app on my phone, so cannot test to send and actual print. Ah maybe I find time to test it with the phone of my wife later 😄.

**Comment by Joulinar** [comment]
Question is if we should rely on Andrew native printing feature or if we can recomend an Android app for users (online docs) 🤔 Maybe I'm thinking to complicate atm

**Comment by MichaIng** [comment]
Usually the native printing provider should work. But if users face issues, or have a custom OS with no printing service pre-installed, we could suggest to try:
- https://f-droid.org/packages/io.github.benoitduffez.cupsprint/
- https://play.google.com/store/apps/details?id=io.github.benoitduffez.cupsprint

Worked fine for me, although there are alternatives as well which I did not test.

**Review by MichaIng — APPROVED** [review]

**Comment by nonnorm** [comment]
It worked for me, but per [this](https://askubuntu.com/questions/82016/network-printer-found-but-cannot-print#answer-455285) askubuntu answer I had to install `libnss-mdns` and change the lines in the config file.  This seems to only be required for Brother printers, but should we install it anyway?
**EDIT**: Notably, this let me `ssh` into my RPi using its `rpi4.local` address, which didn't work before, and this stayed after removing the package, and removing the entry from my `known_hosts` file.
**EDIT2**: Nope, turns out that this it's the Avahi Daemon that let me do this, not the `libnss-mdns` package.

**Comment by Joulinar** [comment]
ok finished testing. However printing on Android is still a mystery for me. Maybe it's me or my devices, I don't know 🤣 It seems to be better on Bullseye. Anyway, ready from my side.

**Comment by MichaIng** [comment]
> EDIT2: Nope, turns out that this it's the Avahi Daemon that let me do this, not the libnss-mdns package.

Good, I was just wondering since one of my successfully tested printers is a brother as well, though an old one 😄. Great I'll merge it then.

**Comment by Joulinar** [comment]
I guess the 2nd update was more related to the fact being able to access the system using `rpi4.local`. the issue with Brother printers seems to be fixed by installing `libnss-mdns` if I'm not mistaken. But this I'm not able to test as I have a HP printer only.

**Comment by MichaIng** [comment]
Hmm, but that indeed should not be related to CUPS at all, as it's only a question whether the hostname.local is made available in the network or not, so indeed the Avahi daemons task, which is pulled as dependency by [libnss-mdns](https://packages.debian.org/buster/libnss-mdns) but installed by `dietpi-software` already.

**Comment by nonnorm** [comment]
Yes, you need `libnss-mdns` for Brother printers to work.  The hostname.local is completely unrelated.

**Comment by MichaIng** [comment]
Strange, in my case if was not required, and I tested it with many different systems, client and host.

... ah I think now I understand in which case it might be required:
- When you add a network printer to CUPS that broadcasts it's own hostname via mDNS (like Avahi daemon), and you want to add it by this mDNS name. @ravenclaw900 does this match your case?
- I had all tested printers locally attached to the host system, respectively passed through to the VM.

I think it's fine to not install it by default but leave a hint in the docs for now. Usually, when you use CUPS, it is not advised to have printers additionally broadcast their own hostname via mDNS, else you would see the same printer two times in the network, one time broadcasted by itself, one time broadcasted by CUPS, isn't it? Probably it's mDNS feature can be disabled and it added via its IP address to CUPS instead to avoid such overhead/collisions.

**Comment by nonnorm** [comment]
Yes, that's what I was testing.  For some reason, that printer wasn't actually showing by default on iOS, and CUPS was needed to make it appear.  However, when I tested with a printer that did appear, it wouldn't produce a duplicate entry.

**Comment by Joulinar** [comment]
I guess there are quite some dependency on local setup and features the printer is providing. If you have a network printer reachable by it's own, do you realy need cups? 😉 

Let's see what people will use it for. I guess it's ok for first attempt and we can adjusted later on.

**Pull request merged**